cmd/create/createCluster.go

```diff
@@ -0,0 +1,419 @@
+
+	// multi-master - datastore
+	// TODO: implement multi-master setups with external data store
+	cmd.Flags().String("datastore-endpoint", "", "[WIP] Specify external datastore endpoint (e.g. for multi master clusters)")
+	/*
+		cmd.Flags().String("datastore-network", "", "Specify container network where we can find the datastore-endpoint (add a connection)")
+
+		// TODO: set default paths and hint, that one should simply mount the files using --volume flag
+		cmd.Flags().String("datastore-cafile", "", "Specify external datastore's TLS Certificate Authority (CA) file")
+		cmd.Flags().String("datastore-certfile", "", "Specify external datastore's TLS certificate file'")
+		cmd.Flags().String("datastore-keyfile", "", "Specify external datastore's TLS key file'")
+	*/
+
+	/* k3s */
+	cmd.Flags().StringArrayVar(&createClusterOpts.K3sServerArgs, "k3s-server-arg", nil, "Additional args passed to the `k3s server` command on master nodes (new flag per arg)")
+	cmd.Flags().StringArrayVar(&createClusterOpts.K3sAgentArgs, "k3s-agent-arg", nil, "Additional args passed to the `k3s agent` command on worker nodes (new flag per arg)")
+
+	/* Subcommands */
+
+	// done
+	return cmd
+}
+
+// parseCreateClusterCmd parses the command input into variables required to create a cluster
+func parseCreateClusterCmd(cmd *cobra.Command, args []string, createClusterOpts *k3d.CreateClusterOpts) *k3d.Cluster {
+
+	/********************************
+	 * Parse and validate arguments *
+	 ********************************/
+
+	clustername := k3d.DefaultClusterName
+	if len(args) != 0 {
+		clustername = args[0]
+	}
+	if err := cluster.CheckName(clustername); err != nil {
+		log.Fatal(err)
+	}
+
+	/****************************
+	 * Parse and validate flags *
+	 ****************************/
+
+	// --image
+	image, err := cmd.Flags().GetString("image")
+	if err != nil {
+		log.Errorln("No image specified")
+		log.Fatalln(err)
+	}
+	if image == "latest" {
+		image = version.GetK3sVersion(true)
+	}
+
+	// --masters
+	masterCount, err := cmd.Flags().GetInt("masters")
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	// --workers
+	workerCount, err := cmd.Flags().GetInt("workers")
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	// --network
+	networkName, err := cmd.Flags().GetString("network")
+	if err != nil {
+		log.Fatalln(err)
+	}
+	network := k3d.ClusterNetwork{}
+	if networkName != "" {
+		network.Name = networkName
+		network.External = true
+	}
+
+	// --secret
+	secret, err := cmd.Flags().GetString("secret")
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	// --wait
+	if cmd.Flags().Changed("wait") && createClusterOpts.WaitForMaster < 0 {
+		log.Fatalln("Value of '--wait' can't be less than 0")
+	}
+
+	// --api-port
+	apiPortFlags, err := cmd.Flags().GetStringArray("api-port")
+	if err != nil {
+		log.Fatalln(err)
+	}
+
+	// error out if we have more api-ports than masters specified
+	if len(apiPortFlags) > masterCount {
```

Comment:
If the user doesn't specify a port can we auto allocated one by looking for a free port based on the current docker containers.

Reply:
What do you mean by "based on the current docker containers"?
At least in a docker environment, we could go "the easy way" of just defining the container port and let docker choose the host port. Then we inspect the created container and draw the chosen port from the spec :thinking: